src/oauth/openidflow.rs

```diff
@@ -0,0 +1,263 @@
+//
+// Copyright 2022 The Sigstore Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! This provides a method for retreiving a OpenID Connect ID Token and scope from the sigstore project.
+//!
+//! The main entry point is the [`OpenIDAuthorize::auth_url`](OpenIDAuthorize::auth_url) function.
+//! This requires four parameters:
+//! - `client_id`: the client ID of the application
+//! - `client_secret`: the client secret of the application
+//! - `issuer`: the URL of the OpenID Connect server
+//! - `redirect_uri`: the URL of the callback endpoint
+//!
+//! The `auth_url` function returns the following:
+//!
+//! - `authorize_url` is a URL that can be opened in a browser. The user will be
+//! prompted to login and authorize the application. The user will be redirected to
+//! the `redirect_uri` URL with a code parameter.
+//!
+//! - `client` is a client object that can be used to make requests to the OpenID
+//! Connect server.
+//!
+//! - `nonce` is a random value that is used to prevent replay attacks.
+//!
+//! - `pkce_verifier` is a PKCE verifier that can be used to generate the code_verifier
+//! value.
+//!
+//! Once you have recieved the above tuple, you can use the [`RedirectListener::redirect_listener`](RedirectListener::redirect_listener)
+//! function to get the ID Token and scope.
+//!
+//! The `redirect_listener` function requires the following parameters:
+//! - `client_redirect_host`: the address for callback.
+//! - `client`: the client object
+//! - `nonce`: the nonce value
+//! - `pkce_verifier`: the PKCE verifier
+//!
+//! The `IdTokenClaims` this contains params such as `email` and the `access_token`.
+//!
+//! It maybe prefered to instead develop your own listener. If so bypass using the
+//! [`RedirectListener::redirect_listener`](RedirectListener::redirect_listener) function and
+//! simply send the values retrieved from the [`OpenIDAuthorize::auth_url`](OpenIDAuthorize::auth_url)
+//! to your own listener.
+//!
+//! **Warning:** the [`OpenIDAuthorize::auth_url`](OpenIDAuthorize::auth_url) function is blocking
+//! and cannot be invoked from within a async function. The  reqwest dependency of oauth2
+//! uses the blocking reqwest feature, e.g:
+//! reqwest = { version = "0.11", optional = true, default-features = false, features = ["blocking"] }
```

Comment:
This is not correct, the function can still be invoked, but with some special handling:

```suggestion
//!
//! **Warning:** one of the dependencies of the [`OpenIDAuthorize::auth_url`](OpenIDAuthorize::auth_url) performs 
//! blocking operations. Because of that it can cause panics at runtime if invoked inside of `async` code.
//! If you need to use this function inside of an async code you must wrap it inside of a `spawn_blocking` instruction:
//!
//! ```
//! use tokio::task::spawn_blocking;
//!
//! async fn my_async_function() {
//!    // ... your code
//!
//!    let oidc_url = spawn_blocking(||
//!     oauth::openidflow::OpenIDAuthorize::new(
//!       "sigstore",
//!       "",
//!       "https://oauth2.sigstore.dev/auth",
//!       "http://localhost:8080",
//!     )
//!     .auth_url()
//!    )
//!    .await
//!    .expect("Error spawning blocking task");
//!
//!    // ... your code
//! }
//! ```
//! This of course has a performance hit when used inside of an async function.
```